### PR TITLE
using mmap dataset for ptuning

### PR DESCRIPTION
**Pull request by arendu**
# What does this PR do ?

replaces standard pytorch dataset with a mem mapped json dataset.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_prompt_learning_dataset.py [hunk 1]
```diff
@@ -60,8 +62,7 @@ def __init__(
         add_eos: bool = True,
         for_train: bool = True,
         tokens_to_generate=None,
-        cache_data_path: str = None,  # the cache file
-        load_cache: bool = True,  # whether to load from the cache if it is available
+        max_num_samples=None,
```
Comment (SeanNaren) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_prompt_learning_dataset.py:65]:
```suggestion
        max_num_samples: Optional[int] = None,
```
Small nit, could we type the ones that do not have types?

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_prompt_learning_dataset.py [hunk 2]
```diff
@@ -245,30 +212,37 @@ def _insert_text_in_template(self, input_example, prompt_template_fields, doc):
...
     def _truncate_input(
-        self, truncation_field, input_ids, taskname, doc, prompt_template, prompt_template_fields, virtual_token_splits
+        self,
+        truncation_field,
+        input_ids,
+        taskname,
+        doc,
+        prompt_template,
+        prompt_template_fields,
+        virtual_token_splits,
```
Comment (SeanNaren) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_prompt_learning_dataset.py:245]: Would be great if we could type these too!

**Review by SeanNaren — DISMISSED** [review]
LGTM! Timely PR was hoping of a way to speedup ptuning on lots of data :)

Just out of curiosity, could we get away without the caching? I was wondering how much a speedup we get from caching as apposed to just relying on the torch dataloader to process everytime.

**Review by yidong72 — COMMENTED** [review]
one minor comment

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_prompt_learning_dataset.py [hunk 3]
```diff
@@ -74,137 +76,105 @@ def __init__(
...
+            taskname_id = self.tokenizer.text_to_ids(taskname)
+        elif self.virtual_prompt_source == VirtualPromptSource.NO_PROMPT:
+            taskname_id = -1
+        else:
+            raise ValueError("Invalid virtual prompt source specified")
+
+        # Find answer field indices if training and answer_only_loss is True
+        answer_start_idx = None
+        if answer_only_loss and self.for_train:
+            answer_start_idx = self._find_answer_start(taskname, input_ids, answer_field, doc)
```
[the remark this answers is not in the document]
Comment (arendu) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_prompt_learning_dataset.py:165]: I have been using this with 43b model which uses sentencepiece and it seems to be ok. But, I do think when sft gets merged, we should switch over to it almost completely.

**Pull request closed without merging**